### PR TITLE
Adding GCLSTM for nodeproppred 

### PR DESCRIPTION
**Pull request by shenyangHuang**
Running into error when running the example
```
python -u gcn.py --dataset tgbn-trade --time-gran r --batch-time-gran r
python -u gcn.py --dataset tgbn-genre --time-gran s --batch-time-gran D
```

- `tgbn-trade` is a yearly trade datasets
`start_time`= 1986 
`end_time`= 2016
`python -u gcn.py --dataset tgbn-trade --time-gran r --batch-time-gran r` 

- `tgbn-genre` is a second wise dataset with snapshots / labels daily
`start_time`= 1108357203
`end_time`= 1245461220
`python -u gcn.py --dataset tgbn-genre --time-gran s --batch-time-gran D` 

- [x] fixed error with not using splits in TGB node prop pred datasets
- [x] added GCLSTM example in nodeproppred achieving `0.4985` NDCG (a few epochs)
- [x] updated GCN example in nodeproppred achieving `0.3164` NDCG (a few epochs)
- [x] fixing error where if some snapshots are missing dynamic node feature, the dataloader wouldn't work

### CODE REVIEW

File: examples/nodeproppred/gcn.py [hunk 1]
```diff
@@ -1,3 +1,8 @@
+r"""python -u gcn.py --dataset tgbn-trade --time-gran r --batch-time-gran r
+python -u gcn.py --dataset tgbn-genre --time-gran s --batch-time-gran D
+example commands to run this script.
+"""
+
```
Comment (Jacob-Chmura) [inline examples/nodeproppred/gcn.py:5]:
```suggestion

```

Just my preference to leave this for docs. Especially since our argparser setup pretty much tells the user how to run things.

File: examples/nodeproppred/gclstm.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        c: torch.Tensor | None,
+    ) -> Tuple[torch.Tensor, ...]:
+        h_0, c_0 = self.recurrent(x, edge_index, edge_weight, h, c)
+        h = F.relu(h_0)
+        h = self.linear(h)
+        return h, h_0, c_0
+
+
+class NodePredictor(torch.nn.Module):
+    def __init__(self, in_dim, out_dim):
```
Comment (Jacob-Chmura) [inline examples/nodeproppred/gclstm.py]:
```suggestion
    def __init__(self, in_dim: int, out_dim: int) -> None:
```

File: tgm/_storage/backends/array_backend.py [hunk 3]
```diff
@@ -159,7 +159,7 @@ def get_dynamic_node_feats(self, slice: DGSliceTracker) -> Optional[Tensor]:
         edge_mask = (self._data.edge_event_idx >= lb_idx) & (
             self._data.edge_event_idx < ub_idx
         )
-        if edge_mask.sum() != 0:
+        if edge_mask.sum() != 0 and self._data.edge_index[edge_mask].numel() > 0:
```
Comment (Jacob-Chmura) [inline tgm/_storage/backends/array_backend.py]:
```suggestion
        if edge_mask.sum() != 0 and len(self._data.edge_index[edge_mask]):
```
I think this is a bit more intuitive, but both are fine

File: tgm/_storage/backends/array_backend.py [hunk 4]
```diff
@@ -188,7 +188,8 @@ def get_edge_feats(self, slice: DGSliceTracker) -> Optional[Tensor]:
             node_mask = (self._data.node_event_idx >= lb_idx) & (
                 self._data.node_event_idx < ub_idx
             )
-            max_node_id = max(max_node_id, self._data.node_ids[node_mask].max())  # type: ignore
+            if self._data.node_ids[node_mask].numel() > 0:  # type: ignore
```
Comment (Jacob-Chmura) [inline tgm/_storage/backends/array_backend.py]:
```suggestion
            if len(self._data.node_ids[node_mask]):  # type: ignore
```
Similarly here

File: tgm/data.py [hunk 5]
```diff
@@ -402,22 +402,25 @@ def _check_tgb_import() -> tuple['LinkPropPredDataset', 'NodePropPredDataset']:
...
-                    temp_node_timestamps[idx] = t
-                    temp_node_ids[idx] = node_id
-                    temp_dynamic_node_feats[idx] = label
-                    idx += 1
+                if (t >= split_start) and (t < split_end):
+                    for node_id, label in node_label_dict[t].items():
+                        temp_node_timestamps[idx] = t
+                        temp_node_ids[idx] = node_id
+                        temp_dynamic_node_feats[idx] = label
+                        idx += 1
```
Comment (Jacob-Chmura) [inline tgm/data.py]: Can we revert this and just do:

```python
node_label_dict = {t: v for t, v in data['node_label_dict'].items() if timestamps[0] <= t < timestamps[-1]
```

File: examples/nodeproppred/gcn.py (hunk 1 above)
Comment (shenyangHuang) [reply examples/nodeproppred/gcn.py:5]: these are specific instructions to make these two dataset work, also ties into their specific instructions, I would leave them here or add them to docs somewhere

File: tgm/data.py (hunk 5 above)
Comment (shenyangHuang) [reply tgm/data.py]: yes
